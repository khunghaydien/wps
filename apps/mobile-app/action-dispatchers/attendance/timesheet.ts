import TimesheetRepository from '../../../repositories/attendance/TimesheetRepository';

export const loadTimesheet = (targetDate: string) =>
  TimesheetRepository.fetch(targetDate);
