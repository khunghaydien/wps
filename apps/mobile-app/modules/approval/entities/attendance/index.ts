import { combineReducers } from 'redux';

import request from './request';

const rootReducer = {
  request,
};

export default combineReducers(rootReducer);
