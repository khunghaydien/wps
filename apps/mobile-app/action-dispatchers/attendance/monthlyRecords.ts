import isNil from 'lodash/isNil';

import { getUserSetting } from '../../../commons/actions/userSetting';
import { catchApiError } from '../../modules/commons/error';
import { withLoading } from '../../modules/commons/loading';

import {
  isTargetDateInTimesheet,
  Timesheet,
} from '../../../domain/models/attendance/Timesheet';

import { actions as EntitiesActions } from '../../modules/attendance/timesheet/entities';
import { actions as UiMonthlyPagingActions } from '../../modules/attendance/timesheet/ui/monthly/paging';

import { AppDispatch } from '../AppThunk';
import { loadTimesheet } from './timesheet';

const clearTimesheet = () => (dispatch: AppDispatch) => {
  dispatch(UiMonthlyPagingActions.clear());
  dispatch(EntitiesActions.clear());
};

const setTimesheet = (timesheet: Timesheet) => (dispatch: AppDispatch) => {
  dispatch(UiMonthlyPagingActions.fetchSuccess(timesheet));
  dispatch(EntitiesActions.fetchSuccess(timesheet));
};

const resetTimesheet =
  (targetDate?: string) =>
  async (dispatch: AppDispatch): Promise<Timesheet | null> => {
    try {
      const timesheet = await loadTimesheet(targetDate || '');
      dispatch(setTimesheet(timesheet));
      return timesheet;
    } catch (error) {
      dispatch(
        catchApiError(error, {
          isContinuable: false,
        })
      );
      dispatch(clearTimesheet());
    }
    return null;
  };

export const initialize =
  (targetDate?: string, timesheet?: Timesheet) =>
  async (dispatch: AppDispatch): Promise<Timesheet | null> => {
    dispatch(clearTimesheet());

    const [_, $timesheet] = await dispatch(
      withLoading(
        (_dispatch: AppDispatch) => _dispatch(getUserSetting()),
        (_dispatch: AppDispatch): any => {
          if (
            !isNil(timesheet) &&
            isTargetDateInTimesheet(timesheet, targetDate)
          ) {
            _dispatch(setTimesheet(timesheet));
            return timesheet;
          } else {
            return _dispatch(resetTimesheet(targetDate));
          }
        }
      )
    );
    return $timesheet;
  };
