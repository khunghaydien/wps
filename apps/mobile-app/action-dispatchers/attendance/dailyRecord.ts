import isNil from 'lodash/isNil';
import moment from 'moment';

import { getUserSetting } from '../../../commons/actions/userSetting';
import msg from '../../../commons/languages';
import { showToast } from '../../../commons/modules/toast';
import TextUtil from '../../../commons/utils/TextUtil';
import { showAlert } from '../../modules/commons/alert';
import { showConfirm } from '../../modules/commons/confirm';
import { catchApiError } from '../../modules/commons/error';
import { withLoading } from '../../modules/commons/loading';

import AttDailyRemarksRepository from '../../../repositories/attendance/AttDailyRemarksRepository';
import AttDailyTimeRepository from '../../../repositories/attendance/AttDailyTimeRepository';
import ManageCommuteCountRepository from '../../../repositories/attendance/ManageCommuteCountRepository';

import { convertToUpdateRequestParam } from '../../../domain/models/attendance/AttDailyTime';
import * as DailyRestTimeFill from '../../../domain/models/attendance/DailyRestTimeFill';
import {
  isTargetDateInTimesheet,
  Timesheet,
} from '../../../domain/models/attendance/Timesheet';

import { actions as EntitiesActions } from '../../modules/attendance/timesheet/entities';
import {
  actions as UiDailyEditingActions,
  State as UiDailyEditingState,
} from '../../modules/attendance/timesheet/ui/daily/editing';
import { actions as UiDailyPagingActions } from '../../modules/attendance/timesheet/ui/daily/paging';

import { AppDispatch } from '../AppThunk';
import { loadTimesheet } from './timesheet';

const clearTimesheet = (targetDate?: string) => (dispatch: AppDispatch) => {
  dispatch(UiDailyPagingActions.clear());
  dispatch(UiDailyEditingActions.clear());
  dispatch(EntitiesActions.clear());
  if (targetDate) {
    dispatch(UiDailyPagingActions.setCurrent(targetDate));
  }
};

const setTimesheet =
  (targetDate: string, timesheet: Timesheet) => (dispatch: AppDispatch) => {
    dispatch(UiDailyEditingActions.fetchSuccess(targetDate, timesheet));
    dispatch(UiDailyPagingActions.fetchSuccess(targetDate, timesheet));
    dispatch(EntitiesActions.fetchSuccess(timesheet));
  };

const resetTimesheet =
  (targetDate: string) =>
  async (dispatch: AppDispatch): Promise<Timesheet | null> => {
    try {
      const timesheet = await loadTimesheet(targetDate);
      dispatch(setTimesheet(targetDate, timesheet));
      return timesheet;
    } catch (error) {
      dispatch(
        catchApiError(error, {
          isContinuable: true,
        })
      );
      dispatch(clearTimesheet(targetDate));
    }
    return null;
  };

export const initialize =
  (_targetDate?: string, timesheet?: Timesheet) =>
  async (dispatch: AppDispatch): Promise<void> => {
    dispatch(clearTimesheet());

    const targetDate = _targetDate || moment().format('YYYY-MM-DD');
    const [_, $timesheet] = await dispatch(
      withLoading(
        (_dispatch: AppDispatch) => _dispatch(getUserSetting()),
        (_dispatch: AppDispatch): any => {
          if (
            !isNil(timesheet) &&
            isTargetDateInTimesheet(timesheet, targetDate)
          ) {
            _dispatch(setTimesheet(targetDate, timesheet));
            return timesheet;
          } else {
            return _dispatch(resetTimesheet(targetDate));
          }
        }
      )
    );

    if ($timesheet && $timesheet.isMigratedSummary) {
      dispatch(showAlert(msg().Att_Err_CanNotDisplayBeforeUsing));
    }
  };

export const saveAttDailyRecord =
  (editing: UiDailyEditingState, useManageCommuteCount) =>
  async (dispatch: AppDispatch) => {
    const { targetDate } = editing;
    await dispatch(
      withLoading(async () => {
        try {
          const result = await AttDailyTimeRepository.update(
            targetDate,
            convertToUpdateRequestParam({
              targetDate,
              startTime: editing.startTime,
              endTime: editing.endTime,
              restTimes: editing.restTimes,
              restHours: editing.restHours,
            })
          );
          await AttDailyRemarksRepository.update({
            recordId: editing.id,
            remarks: editing.remarks || '',
          });

          if (useManageCommuteCount) {
            await ManageCommuteCountRepository.update({
              commuteForwardCount: editing.commuteForwardCount,
              commuteBackwardCount: editing.commuteBackwardCount,
              targetDate,
            });
          }

          if (Number(result.insufficientRestTime)) {
            const confirmText = TextUtil.template(
              msg().Com_Msg_InsufficientRestTime,
              Number(result.insufficientRestTime)
            );
            if (await dispatch(showConfirm(confirmText))) {
              await DailyRestTimeFill.post({ targetDate });
            }
          }
        } catch (error) {
          dispatch(catchApiError(error));
          throw error;
        }
        dispatch(showToast(msg().Att_Lbl_SaveWorkTime));
        await dispatch(resetTimesheet(targetDate));
      })
    );
  };
