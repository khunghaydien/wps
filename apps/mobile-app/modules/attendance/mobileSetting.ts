import { MobileSetting } from '@apps/domain/models/attendance/MobileSetting';

export type State = MobileSetting;

type Initialize = {
  type: 'ATTENDANCE/MOBILE_SETTING';
  payload: State;
};

const ACTIONS = {
  INITIALIZE: 'ATTENDANCE/MOBILE_SETTING',
} as const;

export const actions = {
  initialize: (state: State): Initialize => ({
    type: ACTIONS.INITIALIZE,
    payload: state,
  }),
};

const initialState: State = {
  requireLocationAtMobileStamp: false,
  useManageCommuteCountAtMobileStamp: false,
  defaultCommuteBackwardCount: null,
  defaultCommuteForwardCount: null,
};

type Action = Initialize;

export default (state: State = initialState, action: Action): State => {
  switch (action.type) {
    case ACTIONS.INITIALIZE:
      return action.payload;
    default:
      return state;
  }
};
