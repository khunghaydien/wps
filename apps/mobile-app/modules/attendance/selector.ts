import { AttDailyRequest } from '../../../domain/models/attendance/AttDailyRequest';
import { Code } from '../../../domain/models/attendance/AttDailyRequestType';

import { State } from './index';

export const selectAvailableRequest = (
  state: State,
  requestTypeCode: Code
): AttDailyRequest | null | undefined =>
  state.dailyRequest.entities.availableRequests.find(
    (request) => request.requestTypeCode === requestTypeCode
  ) || null;

export const selectLatestRequest = (
  state: State,
  id: string
): AttDailyRequest | null | undefined =>
  state.dailyRequest.entities.latestRequests.find(
    (request) => request.id === id
  ) || null;
