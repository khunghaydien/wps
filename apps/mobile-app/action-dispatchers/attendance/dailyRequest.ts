import { bindActionCreators, Dispatch } from 'redux';

import isNil from 'lodash/isNil';

import { getUserSetting } from '../../../commons/actions/userSetting';
import msg from '../../../commons/languages';
import { showToast } from '../../../commons/modules/toast';
import { showAlert } from '../../modules/commons/alert';
import { showConfirm } from '../../modules/commons/confirm';
import { catchApiError, catchBusinessError } from '../../modules/commons/error';
import { withLoading } from '../../modules/commons/loading';

import AttDailyLeaveRepository from '../../../repositories/attendance/AttDailyLeaveRepository';
import AttDailyPatternRepository from '../../../repositories/attendance/AttDailyPatternRepository';
import Repository from '../../../repositories/attendance/AttDailyRequestRepository';
import TimesheetRepository from '@apps/repositories/attendance/TimesheetRepository';

import { AttDailyRecord } from '../../../domain/models/attendance/AttDailyRecord';
import {
  AttDailyRequest,
  DISABLE_ACTION,
  EDIT_ACTION,
  isForReapply,
} from '../../../domain/models/attendance/AttDailyRequest';
import * as EarlyLeaveRequest from '../../../domain/models/attendance/AttDailyRequest/EarlyLeaveRequest';
import * as LateArrivalRequest from '../../../domain/models/attendance/AttDailyRequest/LateArrivalRequest';
import * as LeaveRequest from '../../../domain/models/attendance/AttDailyRequest/LeaveRequest';
import * as PatternRequest from '../../../domain/models/attendance/AttDailyRequest/PatternRequest';
import { CODE } from '../../../domain/models/attendance/AttDailyRequestType';
import { doesAttLeaveExist } from '../../../domain/models/attendance/LeaveType';
import {
  getAttDailyRecordByDate,
  isTargetDateInTimesheet,
  Timesheet,
} from '../../../domain/models/attendance/Timesheet';

import { actions as EntitiesAvailableRequestsActions } from '../../modules/attendance/dailyRequest/entities/availableRequests';
import { actions as EntitiesLatestRequestsActions } from '../../modules/attendance/dailyRequest/entities/latestRequests';
import { actions } from '../../modules/attendance/dailyRequest/ui/detail';
import { actions as earlyLeaveRequestActions } from '../../modules/attendance/dailyRequest/ui/requests/earlyLeaveRequest';
import { actions as lateArrivalRequestActions } from '../../modules/attendance/dailyRequest/ui/requests/lateArrivalRequest';
import { actions as leaveRequestActions } from '../../modules/attendance/dailyRequest/ui/requests/leaveRequest';
import { actions as patternRequestActions } from '../../modules/attendance/dailyRequest/ui/requests/patternRequest';
import { actions as ValidationActions } from '../../modules/attendance/dailyRequest/ui/validation';
import {
  actions as TimesheetEntitiesActions,
  State as TimesheetState,
} from '../../modules/attendance/timesheet/entities';
import { actions as TimesheetDailyUiPagingActions } from '../../modules/attendance/timesheet/ui/daily/paging';

import { loadTimesheet } from './timesheet';

const loadAvailableDailyRequest =
  (attDailyRecord: AttDailyRecord, timesheet: Timesheet) =>
  async (dispatch: Dispatch<any>): Promise<void> => {
    const { availableRequestTypeCodesMap } =
      await TimesheetRepository.fetchAvailableDailyRequest(
        attDailyRecord.recordDate
      );
    dispatch(
      EntitiesAvailableRequestsActions.initialize(
        attDailyRecord.id,
        availableRequestTypeCodesMap,
        timesheet
      )
    );
  };

const setDailyRequest =
  (targetDate: string, timesheet: Timesheet) =>
  async (dispatch: Dispatch<any>): Promise<void> => {
    const attDailyRecord = getAttDailyRecordByDate(targetDate, timesheet);
    dispatch(TimesheetDailyUiPagingActions.fetchSuccess(targetDate, timesheet));
    dispatch(TimesheetEntitiesActions.fetchSuccess(timesheet));
    if (attDailyRecord) {
      dispatch(
        EntitiesLatestRequestsActions.initialize(attDailyRecord, timesheet)
      );
      await dispatch(loadAvailableDailyRequest(attDailyRecord, timesheet));
    } else {
      dispatch(EntitiesLatestRequestsActions.clear());
      dispatch(EntitiesAvailableRequestsActions.clear());
    }
  };

const resetDailyRequest =
  (targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<Timesheet | null> => {
    try {
      const timesheet = await loadTimesheet(targetDate);
      await dispatch(setDailyRequest(targetDate, timesheet));
      return timesheet;
    } catch (e) {
      dispatch(
        catchApiError(e, {
          isContinuable: true,
        })
      );
      dispatch(TimesheetDailyUiPagingActions.setCurrent(targetDate));
    }
    return null;
  };

const initializeDailyRequest =
  (targetDate: string, timesheet?: TimesheetState) =>
  async (dispatch: Dispatch<any>) => {
    if (!isNil(timesheet) && isTargetDateInTimesheet(timesheet, targetDate)) {
      await dispatch(setDailyRequest(targetDate, timesheet));
      return timesheet;
    } else {
      return await dispatch(resetDailyRequest(targetDate));
    }
  };

export const initialize =
  (targetDate: string, timesheet?: TimesheetState) =>
  async (dispatch: Dispatch<any>): Promise<void> => {
    dispatch(ValidationActions.clear());
    dispatch(TimesheetDailyUiPagingActions.clear());
    dispatch(TimesheetEntitiesActions.clear());
    dispatch(EntitiesLatestRequestsActions.clear());
    dispatch(EntitiesAvailableRequestsActions.clear());
    // @ts-ignore
    const [_, $timesheet] = await dispatch(
      withLoading(
        // @ts-ignore
        () => dispatch(getUserSetting()),
        () => dispatch(initializeDailyRequest(targetDate, timesheet))
      )
    );
    if ($timesheet && $timesheet.isMigratedSummary) {
      dispatch(showAlert(msg().Att_Err_CanNotDisplayBeforeUsing));
    }
  };

export const startEditing =
  (target: AttDailyRequest, record?: AttDailyRecord | null) =>
  async (dispatch: Dispatch<any>) => {
    dispatch(ValidationActions.clear());
    switch (target.requestTypeCode) {
      case CODE.Leave:
        try {
          await dispatch(
            withLoading(async () => {
              const attLeaveList = await AttDailyLeaveRepository.search({
                targetDate: target.startDate,
                ignoredId: target.id,
              });

              if (attLeaveList.length === 0) {
                dispatch(
                  catchBusinessError(
                    msg().Com_Lbl_Error,
                    msg().Att_Err_CannotEditIfAttLeaveListLengthIsZero,
                    null,
                    { isContinuable: true }
                  )
                );
                return;
              }

              if (isForReapply(target)) {
                if (doesAttLeaveExist(attLeaveList, target)) {
                  const leave = LeaveRequest.create(target, attLeaveList);
                  dispatch(leaveRequestActions.initialize(leave, attLeaveList));
                } else {
                  const leave = LeaveRequest.create(target);
                  dispatch(leaveRequestActions.initialize(leave));
                }
              } else {
                const leave = LeaveRequest.create(target, attLeaveList);
                dispatch(leaveRequestActions.initialize(leave, attLeaveList));
              }
              dispatch(actions.startEditing());
            })
          );
        } catch (e) {
          dispatch(catchApiError(e));
        }

        break;
      case CODE.Pattern:
        try {
          await dispatch(
            withLoading(async () => {
              const attPatternList = await AttDailyPatternRepository.search({
                targetDate: target.startDate,
                ignoredId: target.id,
              });

              if (attPatternList.length === 0) {
                dispatch(
                  catchBusinessError(
                    msg().Com_Lbl_Error,
                    msg().Att_Err_CannotEditIfAttPatternListLengthIsZero,
                    null,
                    { isContinuable: true }
                  )
                );
                return;
              }

              const pattern = PatternRequest.create(
                target,
                attPatternList,
                target.startDate
              );
              dispatch(
                patternRequestActions.initialize(pattern, attPatternList)
              );
              dispatch(actions.startEditing());
            })
          );
        } catch (e) {
          dispatch(catchApiError(e));
        }

        break;
      case CODE.LateArrival:
        const lateArrival = LateArrivalRequest.create(
          target,
          record,
          target.startDate
        );
        dispatch(lateArrivalRequestActions.initialize(lateArrival, record));
        dispatch(actions.startEditing());
        break;
      case CODE.EarlyLeave:
        const earlyLeave = EarlyLeaveRequest.create(
          target,
          record,
          target.startDate
        );
        dispatch(earlyLeaveRequestActions.initialize(earlyLeave, record));
        dispatch(actions.startEditing());
        break;
      default:
        dispatch(actions.startEditing());
        break;
    }
  };

const validate = (target: AttDailyRequest) => (dispatch: Dispatch<any>) =>
  dispatch(ValidationActions.validate(target));

export const create =
  (target: AttDailyRequest, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    const appService = bindActionCreators(
      {
        withLoading,
        catchApiError,
      },
      dispatch
    );
    const dailyRequestService = bindActionCreators(
      { initializeDailyRequest, validate },
      dispatch
    );

    const isValid = await dailyRequestService.validate(target);
    if (!isValid) {
      return false;
    }

    try {
      await appService.withLoading(async () => {
        // submit
        await Repository.create(target);
        await dailyRequestService.initializeDailyRequest(targetDate);
      });
      dispatch(showToast(msg().Att_Lbl_Requested));
      return true;
    } catch (err) {
      appService.catchApiError(err, { isContinuable: true });
      return false;
    }
  };

export const modify =
  (target: AttDailyRequest, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    const appService = bindActionCreators(
      {
        withLoading,
        catchApiError,
      },
      dispatch
    );
    const dailyRequestService = bindActionCreators(
      { initializeDailyRequest, validate },
      dispatch
    );

    const isValid = await dailyRequestService.validate(target);
    if (!isValid) {
      return false;
    }

    try {
      await appService.withLoading(async () => {
        // submit
        await Repository.update(target, EDIT_ACTION.Modify);
        await dailyRequestService.initializeDailyRequest(targetDate);
      });
      dispatch(showToast(msg().Att_Lbl_Requested));
      return true;
    } catch (err) {
      appService.catchApiError(err, { isContinuable: true });
      return false;
    }
  };

export const reapply =
  (target: AttDailyRequest, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    const appService = bindActionCreators(
      {
        withLoading,
        catchApiError,
      },
      dispatch
    );
    const dailyRequestService = bindActionCreators(
      { initializeDailyRequest, validate },
      dispatch
    );

    const isValid = await dailyRequestService.validate(target);
    if (!isValid) {
      return false;
    }

    try {
      await appService.withLoading(async () => {
        await Repository.update(target, EDIT_ACTION.Reapply);
        await dailyRequestService.initializeDailyRequest(targetDate);
      });
      dispatch(showToast(msg().Att_Lbl_Requested));
      return true;
    } catch (err) {
      appService.catchApiError(err, { isContinuable: true });
      return false;
    }
  };

export const cancelRequest =
  (id: string, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    try {
      await dispatch(
        withLoading(async () => {
          await Repository.delete(id, DISABLE_ACTION.CancelRequest);
          await dispatch(initializeDailyRequest(targetDate));
        })
      );
      dispatch(showToast(msg().Att_Lbl_CancelRequest));
      return true;
    } catch (err) {
      dispatch(catchApiError(err, { isContinuable: true }));
      return false;
    }
  };

export const cancelApproval =
  (id: string, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    try {
      await dispatch(
        withLoading(async () => {
          await Repository.delete(id, DISABLE_ACTION.CancelApproval);
          await dispatch(initializeDailyRequest(targetDate));
        })
      );
      dispatch(showToast(msg().Att_Lbl_CancelRequest));
      return true;
    } catch (err) {
      dispatch(catchApiError(err, { isContinuable: true }));
      return false;
    }
  };

export const remove =
  (id: string, targetDate: string) =>
  async (dispatch: Dispatch<any>): Promise<boolean> => {
    if (!(await dispatch(showConfirm(msg().Att_Msg_DailyReqConfirmRemove)))) {
      return false;
    }
    try {
      await dispatch(
        withLoading(async () => {
          await Repository.delete(id, DISABLE_ACTION.Remove);
          await dispatch(initializeDailyRequest(targetDate));
        })
      );
      dispatch(showToast(msg().Att_Lbl_RemoveRequest));
      return true;
    } catch (err) {
      dispatch(catchApiError(err, { isContinuable: true }));
      return false;
    }
  };
