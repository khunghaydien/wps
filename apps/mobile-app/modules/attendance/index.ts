import { combineReducers } from 'redux';

import dailyRequest from './dailyRequest';
import mobileSetting from './mobileSetting';
import timesheet from './timesheet';
import timeStamp from './timeStamp';

const reducers = { timesheet, timeStamp, dailyRequest, mobileSetting };

const rootReducer = combineReducers(reducers);
export type State = ReturnType<typeof rootReducer>;
export default rootReducer;
