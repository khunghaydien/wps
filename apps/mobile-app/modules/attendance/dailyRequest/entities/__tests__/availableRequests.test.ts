import snapshotDiff from 'snapshot-diff';

import { CODE } from '../../../../../../domain/models/attendance/AttDailyRequestType';
import { Timesheet } from '../../../../../../domain/models/attendance/Timesheet';

import reducer, {
  // @ts-ignore
  __get__,
  actions,
} from '../availableRequests';

describe('reducer()', () => {
  it('@@INIT', () => {
    // @ts-ignore
    const next = reducer(undefined, { type: '@@INIT' });
    expect(snapshotDiff([], next)).toMatchSnapshot();
  });

  it(`${__get__('INITIALIZE')}`, () => {
    const map = {
      'request-1': [CODE.Absence, CODE.Direct, CODE.Leave],
      'request-2': [
        CODE.LateArrival,
        CODE.EarlyStartWork,
        CODE.OvertimeWork,
        CODE.EarlyLeave,
      ],
    };
    const timesheet = {
      requestTypes: {
        [CODE.Absence]: {
          code: CODE.Absence,
          name: 'Mock Absence',
        },
        [CODE.Direct]: {
          code: CODE.Direct,
          name: 'Mock Direct',
        },
        [CODE.Leave]: {
          code: CODE.Leave,
          name: 'Mock Leave',
        },
        [CODE.LateArrival]: {
          code: CODE.LateArrival,
          name: 'Mock Late Arrival',
        },
      },
    } as unknown as Timesheet;

    const next = reducer(
      undefined,
      actions.initialize('request-1', map, timesheet)
    );
    expect(snapshotDiff([], next)).toMatchSnapshot();
  });

  it(`${__get__('CLEAR')}`, () => {
    const next = reducer(undefined, actions.clear());
    expect(snapshotDiff([{ requestTypeName: 'test' }], next)).toMatchSnapshot();
  });
});
