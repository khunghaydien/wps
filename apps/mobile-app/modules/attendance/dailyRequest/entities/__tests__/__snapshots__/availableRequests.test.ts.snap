// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer() @@INIT 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`reducer() MOBILE-APP/MODULES/ATTENDANCE/DAILYREQUEST/ENTITIES/AVAILABLEREQUESTS/CLEAR 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Array [
-   Object {
-     \\"requestTypeName\\": \\"test\\",
-   },
- ]
+ Array []"
`;

exports[`reducer() MOBILE-APP/MODULES/ATTENDANCE/DAILYREQUEST/ENTITIES/AVAILABLEREQUESTS/INITIALIZE 1`] = `
"Snapshot Diff:
- First value
+ Second value

- Array []
+ Array [
+   Object {
+     \\"approver01Name\\": \\"\\",
+     \\"directApplyRestTimes\\": Array [],
+     \\"endDate\\": \\"\\",
+     \\"endTime\\": null,
+     \\"id\\": \\"\\",
+     \\"isForReapply\\": false,
+     \\"leaveCode\\": \\"\\",
+     \\"leaveName\\": \\"\\",
+     \\"leaveRange\\": null,
+     \\"leaveType\\": null,
+     \\"originalRequestId\\": \\"\\",
+     \\"patternCode\\": null,
+     \\"patternName\\": null,
+     \\"patternRestTimes\\": Array [],
+     \\"reason\\": \\"\\",
+     \\"remarks\\": \\"\\",
+     \\"requestTypeCode\\": \\"Leave\\",
+     \\"requestTypeName\\": \\"Mock Leave\\",
+     \\"requireReason\\": false,
+     \\"startDate\\": \\"\\",
+     \\"startTime\\": null,
+     \\"status\\": \\"NotRequested\\",
+     \\"substituteDate\\": \\"\\",
+     \\"substituteLeaveType\\": null,
+     \\"type\\": \\"Leave\\",
+   },
+   Object {
+     \\"approver01Name\\": \\"\\",
+     \\"directApplyRestTimes\\": Array [],
+     \\"endDate\\": \\"\\",
+     \\"endTime\\": null,
+     \\"id\\": \\"\\",
+     \\"isForReapply\\": false,
+     \\"leaveCode\\": \\"\\",
+     \\"leaveName\\": \\"\\",
+     \\"leaveRange\\": null,
+     \\"leaveType\\": null,
+     \\"originalRequestId\\": \\"\\",
+     \\"patternCode\\": null,
+     \\"patternName\\": null,
+     \\"patternRestTimes\\": Array [],
+     \\"reason\\": \\"\\",
+     \\"remarks\\": \\"\\",
+     \\"requestTypeCode\\": \\"Direct\\",
+     \\"requestTypeName\\": \\"Mock Direct\\",
+     \\"requireReason\\": false,
+     \\"startDate\\": \\"\\",
+     \\"startTime\\": null,
+     \\"status\\": \\"NotRequested\\",
+     \\"substituteDate\\": \\"\\",
+     \\"substituteLeaveType\\": null,
+     \\"type\\": \\"Direct\\",
+   },
+   Object {
+     \\"approver01Name\\": \\"\\",
+     \\"directApplyRestTimes\\": Array [],
+     \\"endDate\\": \\"\\",
+     \\"endTime\\": null,
+     \\"id\\": \\"\\",
+     \\"isForReapply\\": false,
+     \\"leaveCode\\": \\"\\",
+     \\"leaveName\\": \\"\\",
+     \\"leaveRange\\": null,
+     \\"leaveType\\": null,
+     \\"originalRequestId\\": \\"\\",
+     \\"patternCode\\": null,
+     \\"patternName\\": null,
+     \\"patternRestTimes\\": Array [],
+     \\"reason\\": \\"\\",
+     \\"remarks\\": \\"\\",
+     \\"requestTypeCode\\": \\"Absence\\",
+     \\"requestTypeName\\": \\"Mock Absence\\",
+     \\"requireReason\\": false,
+     \\"startDate\\": \\"\\",
+     \\"startTime\\": null,
+     \\"status\\": \\"NotRequested\\",
+     \\"substituteDate\\": \\"\\",
+     \\"substituteLeaveType\\": null,
+     \\"type\\": \\"Absence\\",
+   },
+ ]"
`;
